### PR TITLE
Fix building SSD index performance issue

### PR DESCRIPTION
**Pull request by yiylin**
- [x] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [x] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
   Add a new parameter to build_merged_vamana_index
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?
 
#### What does this implement/fix? Briefly explain your changes.
There is a big performance regression to build SSD index. Because the num_threads is not passed in, so the method "initialize_query_scratch" can not create enough query_scratch. I applied below one fixing in this PR. Most of the other changing is to fix warnings.

1. Pass num_threads in. This fixing could reduce the building SSD index time from 14023 seconds to 2719 seconds

### CODE REVIEW

**Comment by harsha-simhadri** [comment]
@yiylin Thanks for spotting these errors.

_"There is still a big building index performance gap between SSD and memory because SSD index builds saturate_graph.
If we move below logic after building the whole graph, this could save the building index time a lot."_

_"I could see other places which find a vector(whose time complexity O(n)). we could consider to avoid this or change to use unordered_set"_
Which line of code are you referring to?

_"For the case (num_pq_chunks is equal to the original dimension), we could skip to generate quantized data to save building time and SSD space"_
Would this need a lot of changes to use l2 distance instead of PQ. This is best done in another PR. Could you please post an issue on github so we track it later?

File: tests/utils/compute_groundtruth.cpp [hunk 1]
```diff
@@ -381,9 +381,8 @@ template <typename T>
 int aux_main(const std::string &base_file, const std::string &query_file, const std::string &gt_file, size_t k,
              const diskann::Metric &metric, const std::string &tags_file = std::string(""))
 {
-    size_t npoints, nqueries, dim, npoints_filt;
+    size_t npoints, nqueries, dim;
```
Comment (rakri) [inline tests/utils/compute_groundtruth.cpp:385]: why has base_data gone? was not used is it?

File: src/index.cpp [hunk 2]
```diff
@@ -1388,9 +1402,6 @@ void Index<T, TagT, LabelT>::link(IndexWriteParameters &parameters)
 
     _saturate_graph = parameters.saturate_graph;
 
-    if (num_threads != 0)
```
[the remark this answers is not in the document]
Comment (yiylin) [reply src/index.cpp:1391]: this is a duplicated code in the same method

File: src/index.cpp [hunk 3]
```diff
@@ -1197,6 +1198,10 @@ void Index<T, TagT, LabelT>::occlude_list(const uint32_t location, std::vector<N
         {
             if (occlude_factor[iter - pool.begin()] > cur_alpha)
             {
+                if (skipped_result != nullptr)
+                {
```
[the remark this answers is not in the document]
Comment (yiylin) [reply src/index.cpp:1202]: pruned_list is cleared in line 1298 before calling occlude_list. so "we might have a node both in pruned_list and also in skipped list" can not happen

File: tests/utils/compute_groundtruth.cpp (hunk 1 above)
Comment (yiylin) [reply tests/utils/compute_groundtruth.cpp:385]: yes. they are not used. so I could see the warnings. So I just remove them to fix warnings

File: src/index.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yiylin) [reply src/index.cpp:1202]: yes. you are correct. Let me revert this now. I may try to use the unordered_set in the future.

**Review by harsha-simhadri — APPROVED** [review]

File: src/index.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yiylin) [reply src/index.cpp]: revert this changing since I did not see the gain by using unordered_set

**Pull request merged**